File: Renaissance.AspNet/frontend/Renaissance.Web/Services/HospitalModuleStateService.cs
using Renaissance.Web.Models;

namespace Renaissance.Web.Services;

public sealed class HospitalModuleStateService
{
    private HashSet<AppModule> _enabled = [];
    private HashSet<(AppModule Source, AppModule Target)> _links = [];

    public event Action? Changed;

    public bool IsLoaded { get; private set; }

    public bool IsEnabled(AppModule module)
    {
        if (module is AppModule.Clients or AppModule.ClientDashboard or AppModule.Administration)
        {
            return true;
        }

        return _enabled.Contains(module);
    }

    public IReadOnlyList<AppModule> EnabledModules => _enabled.OrderBy(m => (int)m).ToList();

    public async Task LoadAsync(RenaissanceApiClient api, CancellationToken ct = default)
    {
        var state = await api.GetActiveHospitalModulesAsync(ct);
        Apply(state);
    }

    public void Apply(HospitalModulesStateDto state)
    {
        _enabled = state.EnabledModules.ToHashSet();
        _links = state.ReferralLinks
            .Select(l => (l.SourceModule, l.TargetModule))
            .ToHashSet();
        IsLoaded = true;
        Changed?.Invoke();
    }

    public bool CanRefer(AppModule source, AppModule target)
        => source != target && _links.Contains((source, target));

    public IEnumerable<ModuleNavItem> GetReferralTargets(AppModule source)
        => ModuleNav.ReferralTargets
            .Where(item => CanRefer(source, item.Module));

    public IEnumerable<ModuleNavItem> FilterNavItems(IEnumerable<ModuleNavItem> items)
        => items.Where(item => IsEnabled(item.Module));
}

File: Renaissance.AspNet/frontend/Renaissance.Angular/src/app/core/services/hospital-module-state.service.ts
import { Injectable } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { AppModule, ModuleNavItem, MODULE_NAV_REFERRAL_TARGETS } from '../models/auth.models';
import { HospitalModulesStateDto } from '../models/settings.models';

@Injectable({ providedIn: 'root' })
export class HospitalModuleStateService {
    private enabled = new Set<AppModule>();
    private referralLinks = new Set<string>();
    private readonly _loaded$ = new BehaviorSubject<boolean>(false);

    readonly loaded$ = this._loaded$.asObservable();

    isEnabled(module: AppModule): boolean {
        if (module === AppModule.Clients || module === AppModule.ClientDashboard || module === AppModule.Administration) {
            return true;
        }

        return this.enabled.has(module);
    }

    apply(state: HospitalModulesStateDto): void {
        this.enabled = new Set(state.enabledModules);
        this.referralLinks = new Set(
            state.referralLinks.map(link => `${link.sourceModule}-${link.targetModule}`)
        );
        this._loaded$.next(true);
    }

    canRefer(source: AppModule, target: AppModule): boolean {
        return source !== target && this.referralLinks.has(`${source}-${target}`);
    }

    getReferralTargets(source: AppModule): ModuleNavItem[] {
        return MODULE_NAV_REFERRAL_TARGETS.filter(item => this.canRefer(source, item.module));
    }

    filterNavItems(items: ModuleNavItem[]): ModuleNavItem[] {
        return items.filter(item => this.isEnabled(item.module));
    }
}
